File: src/core/event_msg.cpp
#include "event_msg.h"
#include "../utils/debug.h"
#include <map>

// ═══════════════════════════════════════════════════════
// INTERNAL STATE
// ═══════════════════════════════════════════════════════

// Decoder state machine states
enum DecoderState {
    STATE_IDLE,          // Waiting for SOH
    STATE_WAIT_STX,      // Waiting for STX after header (skips header bytes)
    STATE_READ_NAME,     // Reading event name
    STATE_READ_DATA,     // Reading event data
    STATE_ESCAPE         // Next byte is escaped
};

// Decoder state
static DecoderState decoder_state = STATE_IDLE;
static String event_name = "";
static std::vector<uint8_t> event_data;  // Dynamic - no size limit
static bool in_name_section = true;

// Encoder state (for auto-incrementing message ID)
static uint16_t next_message_id = 0;

// Event handler registry
static std::map<String, EventHandler> event_handlers;
static UnhandledEventHandler unhandled_handler = nullptr;

// Send callback
static EventSendCallback send_callback = nullptr;

// ═══════════════════════════════════════════════════════
// ENCODER (with byte stuffing)
// ═══════════════════════════════════════════════════════

// Helper: Check if byte needs stuffing
static bool needs_stuffing(uint8_t byte) {
    return (byte == MSG_SOH ||byte == MSG_STX || byte == MSG_US || byte == MSG_EOT || byte == MSG_ESC );
}

// Helper: Stuff a single byte into buffer
static uint16_t stuff_byte(uint8_t byte, uint8_t* buffer, uint16_t index) {
    if (needs_stuffing(byte)) {
        buffer[index++] = MSG_ESC;
        buffer[index++] = byte ^ MSG_ESC_XOR;
    } else {
        buffer[index++] = byte;
    }
    return index;
}

// Helper: Create and byte-stuff header
// Returns number of bytes written
static uint16_t create_header(uint8_t* buffer, uint16_t index,
                               uint8_t sender_id, uint8_t receiver_id,
                               uint8_t sender_group_id, uint8_t receiver_group_id,
                               uint8_t flags, uint16_t message_id) {
    uint16_t start_idx = index;
    uint16_t stuffed_count = 0;

    LOG_TRACE("ENCODE", "Creating header: sender=%d, receiver=%d, groups=%d/%d, flags=0x%02X, msgID=%d",
              sender_id, receiver_id, sender_group_id, receiver_group_id, flags, message_id);

    // Stuff senderId
    uint16_t old_idx = index;
    index = stuff_byte(sender_id, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    // Stuff receiverId
    old_idx = index;
    index = stuff_byte(receiver_id, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    // Stuff senderGroupId
    old_idx = index;
    index = stuff_byte(sender_group_id, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    // Stuff receiverGroupId
    old_idx = index;
    index = stuff_byte(receiver_group_id, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    // Stuff flags
    old_idx = index;
    index = stuff_byte(flags, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    // Stuff messageId (16-bit, big-endian MSB first)
    uint8_t msg_id_msb = (message_id >> 8) & 0xFF;
    uint8_t msg_id_lsb = message_id & 0xFF;

    old_idx = index;
    index = stuff_byte(msg_id_msb, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    old_idx = index;
    index = stuff_byte(msg_id_lsb, buffer, index);
    if (index - old_idx > 1) stuffed_count++;

    LOG_TRACE("ENCODE", "  Header created: %d bytes (7 logical, %d stuffed)",
              index - start_idx, stuffed_count);

    return index;
}

uint16_t event_msg_encode(const char* name, const uint8_t* data, uint16_t data_len, uint8_t* out_buffer) {
    uint16_t idx = 0;

    LOG_TRACE("ENCODE", "Encoding event '%s' with %d bytes of data", name, data_len);

    // Frame start - SOH
    out_buffer[idx++] = MSG_SOH;
    LOG_TRACE("ENCODE", "  [%d] SOH (0x%02X)", idx-1, MSG_SOH);

    // Create and byte-stuff header
    uint16_t header_start = idx;
    idx = create_header(out_buffer, idx,
                        1,              // senderId
                        0,              // receiverId
                        0,              // senderGroupId
                        0,              // receiverGroupId
                        0,              // flags
                        next_message_id // messageId
    );
    LOG_TRACE("ENCODE", "  [%d-%d] Header stuffed (%d bytes, msgID=%d)",
              header_start, idx-1, idx - header_start, next_message_id);

    // Increment message ID for next send
    next_message_id++;

    // Start of text
    out_buffer[idx++] = MSG_STX;
    LOG_TRACE("ENCODE", "  [%d] STX (0x%02X)", idx-1, MSG_STX);

    // Stuff event name
    uint16_t name_start = idx;
    uint16_t name_len = 0;
    uint16_t name_stuffed = 0;
    for (uint16_t i = 0; name[i] != '\0'; i++) {
        name_len++;
        uint16_t old_idx = idx;
        idx = stuff_byte(name[i], out_buffer, idx);
        if (idx - old_idx > 1) {
            name_stuffed++;
            LOG_TRACE("ENCODE", "    Name byte stuffed: 0x%02X -> ESC+0x%02X", name[i], name[i] ^ MSG_ESC_XOR);
        }
    }
    LOG_TRACE("ENCODE", "  [%d-%d] Event name '%s' (%d chars, %d stuffed)",
              name_start, idx-1, name, name_len, name_stuffed);

    // Unit separator
    out_buffer[idx++] = MSG_US;
    LOG_TRACE("ENCODE", "  [%d] US (0x%02X)", idx-1, MSG_US);

    // Stuff event data
    uint16_t data_start = idx;
    uint16_t data_stuffed = 0;
    for (uint16_t i = 0; i < data_len; i++) {
        uint16_t old_idx = idx;
        idx = stuff_byte(data[i], out_buffer, idx);
        if (idx - old_idx > 1) {
            data_stuffed++;
        }
    }
    LOG_TRACE("ENCODE", "  [%d-%d] Event data (%d bytes, %d stuffed)",
              data_start, idx-1, data_len, data_stuffed);

    // Frame end
    out_buffer[idx++] = MSG_EOT;
    LOG_TRACE("ENCODE", "  [%d] EOT (0x%02X)", idx-1, MSG_EOT);

    LOG_DEBUG("ENCODE", "Encoded '%s': %d bytes total (name=%d, data=%d, stuffed=%d)",
              name, idx, name_len, data_len, name_stuffed + data_stuffed);

    return idx;  // Return total encoded length
}

// ═══════════════════════════════════════════════════════
// PUBLIC API
// ═══════════════════════════════════════════════════════

void event_msg_init(EventSendCallback on_send) {
    LOG_INFO("EVENT", "Initializing event message system");
    send_callback = on_send;
    decoder_state = STATE_IDLE;
    event_name = "";
    event_data.clear();
    in_name_section = true;
    next_message_id = 0;
    event_handlers.clear();
    unhandled_handler = nullptr;
    LOG_DEBUG("EVENT", "Event system initialized - decoder ready");
}

void event_msg_on(const char* event_name, EventHandler handler) {
    if (handler != nullptr) {
        event_handlers[String(event_name)] = handler;
        LOG_DEBUG("EVENT", "Registered handler for '%s'", event_name);
    }
}

void event_msg_on_unhandled(UnhandledEventHandler handler) {
    unhandled_handler = handler;
    LOG_DEBUG("EVENT", "Registered unhandled event handler");
}

void event_msg_send(const char* name, const uint8_t* data, uint16_t len) {
    if (send_callback == nullptr) {
        LOG_ERROR("EVENT", "Cannot send '%s' - no send callback registered", name);
        return;  // No send callback registered
    }

    LOG_DEBUG("EVENT", "Sending event '%s' with %d bytes", name, len);

    // Use static buffer to avoid stack overflow (ESP32 has limited stack)
    static uint8_t buffer[4*1024];  // Static allocation - not on stack
    uint16_t encoded_len = event_msg_encode(name, data, len, buffer);

    LOG_DEBUG("EVENT", "Calling send callback with %d encoded bytes", encoded_len);

    // Send via callback
    send_callback(buffer, encoded_len);

    LOG_TRACE("EVENT", "Event '%s' sent successfully", name);
}

void event_msg_feed_bytes(const uint8_t* data, uint16_t len) {
    LOG_DEBUG("DECODE", "Feeding %d bytes to decoder", len);
    for (uint16_t i = 0; i < len; i++) {
        event_msg_feed_byte(data[i]);
    }
}

void event_msg_reset() {
    LOG_DEBUG("DECODE", "Decoder reset - returning to IDLE state");
    decoder_state = STATE_IDLE;
    event_name = "";
    event_data.clear();
    in_name_section = true;
}

void event_msg_feed_byte(uint8_t byte) {
    LOG_TRACE("DECODE", "RX byte: 0x%02X (state=%d)", byte, decoder_state);

    switch (decoder_state) {

        // ─────────────────────────────────────────────────
        // IDLE: Waiting for frame start (SOH)
        // ─────────────────────────────────────────────────
        case STATE_IDLE:
            if (byte == MSG_SOH) {
                // Frame started - now wait for STX (skip variable-length header)
                LOG_DEBUG("DECODE", "Frame start (SOH) - waiting for STX");
                decoder_state = STATE_WAIT_STX;
            } else {
                LOG_TRACE("DECODE", "Ignoring byte 0x%02X (waiting for SOH)", byte);
            }
            break;

        // ─────────────────────────────────────────────────
        // WAIT_STX: Waiting for STX after header (header may be byte-stuffed)
        // ─────────────────────────────────────────────────
        case STATE_WAIT_STX:
            if (byte == MSG_STX) {
                // STX found - reset state and start reading name
                LOG_DEBUG("DECODE", "STX found - reading event name");
                event_name = "";
                event_data.clear();
                in_name_section = true;
                decoder_state = STATE_READ_NAME;
            } else if (byte == MSG_SOH) {
                // New frame started - restart
                LOG_DEBUG("DECODE", "New SOH while waiting for STX - restarting");
                decoder_state = STATE_WAIT_STX;
            } else {
                LOG_TRACE("DECODE", "Skipping header byte: 0x%02X", byte);
            }
            // Otherwise, keep waiting for STX (skip all header bytes)
            break;

        // ─────────────────────────────────────────────────
        // READ_NAME: Reading event name until US
        // ─────────────────────────────────────────────────
        case STATE_READ_NAME:
            if (byte == MSG_ESC) {
                // Next byte is escaped
                LOG_TRACE("DECODE", "ESC in name - next byte will be unstuffed");
                decoder_state = STATE_ESCAPE;
            } else if (byte == MSG_US) {
                // Name complete, start reading data
                LOG_DEBUG("DECODE", "Event name complete: '%s' - reading data", event_name.c_str());
                in_name_section = false;
                decoder_state = STATE_READ_DATA;
            } else if (byte == MSG_SOH) {
                // New frame started - restart from beginning
                LOG_DEBUG("DECODE", "SOH during name read - frame aborted, restarting");
                decoder_state = STATE_WAIT_STX;
            } else if (byte == MSG_STX) {
                // Unexpected STX - restart frame parsing (already past header)
                LOG_DEBUG("DECODE", "Unexpected STX during name read - restarting frame");
                event_name = "";
                event_data.clear();
                in_name_section = true;
                decoder_state = STATE_READ_NAME;
            } else {
                // Normal character - add to name
                event_name += (char)byte;
                LOG_TRACE("DECODE", "Name char: '%c' (0x%02X) - name so far: '%s'",
                          (char)byte, byte, event_name.c_str());
            }
            break;

        // ─────────────────────────────────────────────────
        // READ_DATA: Reading event data until EOT
        // ─────────────────────────────────────────────────
        case STATE_READ_DATA:
            if (byte == MSG_ESC) {
                // Next byte is escaped
                LOG_TRACE("DECODE", "ESC in data - next byte will be unstuffed");
                decoder_state = STATE_ESCAPE;
            } else if (byte == MSG_EOT) {
                // Frame complete - dispatch to registered handler
                LOG_DEBUG("DECODE", "Frame complete: '%s' with %d bytes of data",
                          event_name.c_str(), event_data.size());

                auto it = event_handlers.find(event_name);
                if (it != event_handlers.end()) {
                    // Handler found - call it
                    LOG_DEBUG("EVENT", "Dispatching '%s' to handler (%d bytes)",
                              event_name.c_str(), event_data.size());
                    it->second(event_data);
                } else if (unhandled_handler != nullptr) {
                    // No specific handler, call unhandled handler with event name
                    LOG_DEBUG("EVENT", "Unhandled event '%s', calling wildcard handler",
                              event_name.c_str());
                    unhandled_handler(event_name, event_data);
                } else {
                    // No handler registered
                    LOG_DEBUG("EVENT", "Warning: No handler for '%s'", event_name.c_str());
                }
                // Reset for next frame
                event_msg_reset();
            } else if (byte == MSG_SOH) {
                // New frame started - restart from beginning
                LOG_DEBUG("DECODE", "SOH during data read - frame aborted, restarting");
                decoder_state = STATE_WAIT_STX;
            } else if (byte == MSG_STX) {
                // Unexpected STX - restart frame parsing (already past header)
                LOG_DEBUG("DECODE", "Unexpected STX during data read - restarting frame");
                event_name = "";
                event_data.clear();
                in_name_section = true;
                decoder_state = STATE_READ_NAME;
            } else {
                // Normal data byte - add to vector (no size limit)
                event_data.push_back(byte);
                LOG_TRACE("DECODE", "Data byte: 0x%02X (%d bytes total)", byte, event_data.size());
            }
            break;

        // ─────────────────────────────────────────────────
        // ESCAPE: Unstuff next byte
        // ─────────────────────────────────────────────────
        case STATE_ESCAPE:
            {
                uint8_t unstuffed = byte ^ MSG_ESC_XOR;

                if (in_name_section) {
                    // Add unstuffed byte to name
                    event_name += (char)unstuffed;
                    LOG_TRACE("DECODE", "Unstuffed name byte: 0x%02X -> '%c' (0x%02X)",
                              byte, (char)unstuffed, unstuffed);
                    decoder_state = STATE_READ_NAME;
                } else {
                    // Add unstuffed byte to data (no size limit)
                    event_data.push_back(unstuffed);
                    LOG_TRACE("DECODE", "Unstuffed data byte: 0x%02X -> 0x%02X",
                              byte, unstuffed);
                    decoder_state = STATE_READ_DATA;
                }
            }
            break;
    }
}

File: web/eventstudio/event_msg.js
/**
 * EVENT MESSAGE PROTOCOL - JavaScript Implementation
 *
 * Binary-safe event messaging protocol with byte stuffing
 * Compatible with ESP32 event_msg.cpp implementation
 *
 * Frame Format:
 *   [SOH] [7-byte header] [STX] [Stuffed Event Name] [US] [Stuffed Event Data] [EOT]
 *
 * Header Format (7 bytes):
 *   senderId (1 byte)
 *   receiverId (1 byte)
 *   senderGroupId (1 byte)
 *   receiverGroupId (1 byte)
 *   flags (1 byte)
 *   messageId (2 bytes, big-endian MSB first)
 *
 * Control Characters:
 *   SOH  = 0x01 (Start of Header)
 *   STX  = 0x02 (Start of Text)
 *   US   = 0x1F (Unit Separator - between name and data)
 *   EOT  = 0x04 (End of Transmission)
 *   ESC  = 0x1B (Escape for byte stuffing)
 *
 * Byte Stuffing:
 *   If payload contains STX, US, EOT, or ESC:
 *     Send: [ESC][byte XOR 0x20]
 *   On receive:
 *     If ESC found: Next byte XOR 0x20
 */

// ═══════════════════════════════════════════════════════
// CONTROL CHARACTERS
// ═══════════════════════════════════════════════════════

const MSG_SOH = 0x01;      // Start of Header
const MSG_STX = 0x02;      // Start of Text
const MSG_US = 0x1F;       // Unit Separator
const MSG_EOT = 0x04;      // End of Transmission
const MSG_ESC = 0x1B;      // Escape
const MSG_ESC_XOR = 0x20;  // XOR mask for stuffing

// Header constants
const MSG_HEADER_SIZE = 7;

// Global message ID counter (auto-increments)
let nextMessageId = 0;

// ═══════════════════════════════════════════════════════
// ENCODER (with byte stuffing)
// ═══════════════════════════════════════════════════════

/**
 * Check if byte needs stuffing
 */
function needsStuffing(byte) {
    return (byte === MSG_STX || byte === MSG_US || byte === MSG_EOT || byte === MSG_ESC);
}

/**
 * Stuff a single byte into output array
 */
function stuffByte(byte, output) {
    if (needsStuffing(byte)) {
        output.push(MSG_ESC);
        output.push(byte ^ MSG_ESC_XOR);
    } else {
        output.push(byte);
    }
}

/**
 * Encode an event into a byte array
 * @param {string} name - Event name
 * @param {Uint8Array|Array|string} data - Event data (can be binary or string)
 * @returns {Uint8Array} - Encoded frame
 */
export function encodeEvent(name, data = null) {
    const output = [];

    // Frame start - SOH
    output.push(MSG_SOH);

    // 7-byte header (fixed values)
    output.push(1);  // senderId
    output.push(0);  // receiverId
    output.push(0);  // senderGroupId
    output.push(0);  // receiverGroupId
    output.push(0);  // flags

    // messageId (16-bit, big-endian MSB first)
    output.push((nextMessageId >> 8) & 0xFF);  // MSB
    output.push(nextMessageId & 0xFF);         // LSB

    // Increment message ID for next send
    nextMessageId = (nextMessageId + 1) & 0xFFFF;  // Wrap at 16-bit

    // Start of text
    output.push(MSG_STX);

    // Stuff event name (convert string to bytes)
    for (let i = 0; i < name.length; i++) {
        stuffByte(name.charCodeAt(i), output);
    }

    // Unit separator
    output.push(MSG_US);

    // Stuff event data
    if (data !== null) {
        // Handle different data types
        let dataBytes;
        if (typeof data === 'string') {
            // Convert string to UTF-8 bytes
            dataBytes = new TextEncoder().encode(data);
        } else if (data instanceof Uint8Array) {
            dataBytes = data;
        } else if (Array.isArray(data)) {
            dataBytes = new Uint8Array(data);
        } else {
            dataBytes = new Uint8Array(0);
        }

        for (let i = 0; i < dataBytes.length; i++) {
            stuffByte(dataBytes[i], output);
        }
    }

    // Frame end
    output.push(MSG_EOT);

    return new Uint8Array(output);
}

// ═══════════════════════════════════════════════════════
// DECODER (with state machine)
// ═══════════════════════════════════════════════════════

const DecoderState = {
    IDLE: 0,          // Waiting for SOH
    SKIP_HEADER: 1,   // Skipping 7-byte header
    WAIT_STX: 2,      // Waiting for STX after header
    READ_NAME: 3,     // Reading event name
    READ_DATA: 4,     // Reading event data
    ESCAPE: 5         // Next byte is escaped
};

export class EventDecoder {
    constructor() {
        this.state = DecoderState.IDLE;
        this.eventName = '';
        this.eventData = [];
        this.inNameSection = true;
        this.headerBytesRead = 0;
        this.eventHandlers = new Map();
        this.unhandledHandler = null;
    }

    /**
     * Register handler for specific event name
     * @param {string} eventName - Name of the event
     * @param {Function} handler - Callback function(data)
     */
    on(eventName, handler) {
        this.eventHandlers.set(eventName, handler);
        console.log(`[EVENT] Registered handler for '${eventName}'`);
    }

    /**
     * Set wildcard handler for unhandled events
     * @param {Function} handler - Callback function(name, data)
     */
    onUnhandled(handler) {
        this.unhandledHandler = handler;
        console.log('[EVENT] Registered unhandled event handler');
    }

    /**
     * Reset decoder state
     */
    reset() {
        this.state = DecoderState.IDLE;
        this.eventName = '';
        this.eventData = [];
        this.inNameSection = true;
        this.headerBytesRead = 0;
    }

    /**
     * Feed a single byte to the decoder
     * @param {number} byte - Byte to process
     */
    feedByte(byte) {
        switch (this.state) {

            // ─────────────────────────────────────────────────
            // IDLE: Waiting for frame start (SOH)
            // ─────────────────────────────────────────────────
            case DecoderState.IDLE:
                if (byte === MSG_SOH) {
                    // Frame started - prepare to skip header
                    this.headerBytesRead = 0;
                    this.state = DecoderState.SKIP_HEADER;
                }
                break;

            // ─────────────────────────────────────────────────
            // SKIP_HEADER: Skip 7-byte header
            // ─────────────────────────────────────────────────
            case DecoderState.SKIP_HEADER:
                this.headerBytesRead++;
                if (this.headerBytesRead >= MSG_HEADER_SIZE) {
                    // All 7 header bytes skipped, now wait for STX
                    this.state = DecoderState.WAIT_STX;
                }
                break;

            // ─────────────────────────────────────────────────
            // WAIT_STX: Waiting for STX after header
            // ─────────────────────────────────────────────────
            case DecoderState.WAIT_STX:
                if (byte === MSG_STX) {
                    // STX found - reset state and start reading name
                    this.eventName = '';
                    this.eventData = [];
                    this.inNameSection = true;
                    this.state = DecoderState.READ_NAME;
                } else if (byte === MSG_SOH) {
                    // New frame started - restart
                    this.headerBytesRead = 0;
                    this.state = DecoderState.SKIP_HEADER;
                } else {
                    // Unexpected byte - reset to idle
                    this.reset();
                }
                break;

            // ─────────────────────────────────────────────────
            // READ_NAME: Reading event name until US
            // ─────────────────────────────────────────────────
            case DecoderState.READ_NAME:
                if (byte === MSG_ESC) {
                    // Next byte is escaped
                    this.state = DecoderState.ESCAPE;
                } else if (byte === MSG_US) {
                    // Name complete, start reading data
                    this.inNameSection = false;
                    this.state = DecoderState.READ_DATA;
                } else if (byte === MSG_SOH) {
                    // New frame started - restart from beginning
                    this.headerBytesRead = 0;
                    this.state = DecoderState.SKIP_HEADER;
                } else if (byte === MSG_STX) {
                    // Unexpected STX - restart frame parsing (already past header)
                    this.eventName = '';
                    this.eventData = [];
                    this.inNameSection = true;
                    this.state = DecoderState.READ_NAME;
                } else {
                    // Normal character - add to name
                    this.eventName += String.fromCharCode(byte);
                }
                break;

            // ─────────────────────────────────────────────────
            // READ_DATA: Reading event data until EOT
            // ─────────────────────────────────────────────────
            case DecoderState.READ_DATA:
                if (byte === MSG_ESC) {
                    // Next byte is escaped
                    this.state = DecoderState.ESCAPE;
                } else if (byte === MSG_EOT) {
                    // Frame complete - dispatch to registered handler
                    this._dispatchEvent();
                    this.reset();
                } else if (byte === MSG_SOH) {
                    // New frame started - restart from beginning
                    this.headerBytesRead = 0;
                    this.state = DecoderState.SKIP_HEADER;
                } else if (byte === MSG_STX) {
                    // Unexpected STX - restart frame parsing (already past header)
                    this.eventName = '';
                    this.eventData = [];
                    this.inNameSection = true;
                    this.state = DecoderState.READ_NAME;
                } else {
                    // Normal data byte - add to data
                    this.eventData.push(byte);
                }
                break;

            // ─────────────────────────────────────────────────
            // ESCAPE: Unstuff next byte
            // ─────────────────────────────────────────────────
            case DecoderState.ESCAPE:
                {
                    const unstuffed = byte ^ MSG_ESC_XOR;

                    if (this.inNameSection) {
                        // Add unstuffed byte to name
                        this.eventName += String.fromCharCode(unstuffed);
                        this.state = DecoderState.READ_NAME;
                    } else {
                        // Add unstuffed byte to data
                        this.eventData.push(unstuffed);
                        this.state = DecoderState.READ_DATA;
                    }
                }
                break;
        }
    }

    /**
     * Feed multiple bytes to the decoder
     * @param {Uint8Array|Array} bytes - Bytes to process
     */
    feedBytes(bytes) {
        for (let i = 0; i < bytes.length; i++) {
            this.feedByte(bytes[i]);
        }
    }

    /**
     * Dispatch event to registered handler (internal)
     */
    _dispatchEvent() {
        const dataArray = new Uint8Array(this.eventData);

        if (this.eventHandlers.has(this.eventName)) {
            // Handler found - call it
            console.log(`[EVENT] Dispatching '${this.eventName}' (${dataArray.length} bytes)`);
            const handler = this.eventHandlers.get(this.eventName);
            handler(dataArray);
        } else if (this.unhandledHandler !== null) {
            // No specific handler, call unhandled handler with event name
            console.log(`[EVENT] Unhandled event '${this.eventName}', calling wildcard handler`);
            this.unhandledHandler(this.eventName, dataArray);
        } else {
            // No handler registered
            console.log(`[EVENT] Warning: No handler for '${this.eventName}'`);
        }
    }
}

// ═══════════════════════════════════════════════════════
// HELPER UTILITIES
// ═══════════════════════════════════════════════════════

/**
 * Convert Uint8Array to hex string for debugging
 * @param {Uint8Array} data - Data to convert
 * @returns {string} - Hex string
 */
export function toHexString(data) {
    return Array.from(data)
        .map(b => b.toString(16).padStart(2, '0').toUpperCase())
        .join(' ');
}

/**
 * Convert Uint8Array to UTF-8 string
 * @param {Uint8Array} data - Data to convert
 * @returns {string} - UTF-8 string
 */
export function toString(data) {
    return new TextDecoder().decode(data);
}

/**
 * Convert hex string to Uint8Array
 * @param {string} hexString - Hex string (e.g., "01 02 03" or "010203")
 * @returns {Uint8Array} - Byte array
 */
export function fromHexString(hexString) {
    const hex = hexString.replace(/\s+/g, '');
    const bytes = [];
    for (let i = 0; i < hex.length; i += 2) {
        bytes.push(parseInt(hex.substr(i, 2), 16));
    }
    return new Uint8Array(bytes);
}
